simulator/src/main/java/com/hazelcast/simulator/protocol/core/MessageFuture.java

```diff
@@ -0,0 +1,81 @@
+package com.hazelcast.simulator.protocol.core;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static java.lang.String.format;
+
+public final class MessageFuture<E> implements Future<E> {
+
+    private static final Object NO_RESULT = new Object();
+
+    @SuppressWarnings("unchecked")
+    private volatile E result = (E) NO_RESULT;
+
+    public MessageFuture() {
+    }
+
+    @Override
+    public boolean cancel(boolean mayInterruptIfRunning) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean isCancelled() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean isDone() {
+        return (result != NO_RESULT);
+    }
+
+    public void set(E result) {
+        synchronized (this) {
+            this.result = result;
+            notifyAll();
+        }
+    }
+
+    @Override
+    public E get() throws InterruptedException, ExecutionException {
+        synchronized (this) {
+            while (NO_RESULT.equals(result)) {
+                wait();
+            }
+
+            if (result instanceof Throwable) {
+                throw new ExecutionException((Throwable) result);
+            }
+            return result;
+        }
+    }
+
+    @Override
+    public E get(long timeout, TimeUnit timeUnit) throws InterruptedException, ExecutionException, TimeoutException {
+        if (timeout < 0 || timeUnit == null) {
+            throw new IllegalArgumentException("Invalid timeout or timeUnit for MessageFuture.get()");
+        }
+
+        long remainingTimeoutNanos = timeUnit.toNanos(timeout);
+        synchronized (this) {
+            while (NO_RESULT.equals(result) && remainingTimeoutNanos > 0) {
+                long started = System.nanoTime();
+                wait(TimeUnit.NANOSECONDS.toMillis(remainingTimeoutNanos));
```

Comment:
can't this lead to 0 due rounding errors -> `wait()` will treat it as indefinite timeout?